#**EDA For Global COVID-19 Cases Dashboard**

##**Import Package**


In [4]:
#importing package from libraries
import warnings
warnings.filterwarnings("ignore") #for ignore not important warnings
import numpy as np #for saintific process
import pandas as pd #for dataframe processing
import seaborn as sns #for data visualitation
import matplotlib.pyplot as plt #for data visualitation
%matplotlib inline

In [5]:
from google.colab import drive #for connect to drive
drive.mount('/content/drive')

Mounted at /content/drive


##**Data Understading**

###1.Data Identification

In [6]:
#load data train 
pd.set_option('display.max_columns', 200)
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/train_global_covid.csv')
df.head()

Id County Province_State Country_Region  Population    Weight        Date  \
0   1    NaN            NaN    Afghanistan    27657145  0.058359  2020-01-23   
1   2    NaN            NaN    Afghanistan    27657145  0.583587  2020-01-23   
2   3    NaN            NaN    Afghanistan    27657145  0.058359  2020-01-24   
3   4    NaN            NaN    Afghanistan    27657145  0.583587  2020-01-24   
4   5    NaN            NaN    Afghanistan    27657145  0.058359  2020-01-25   

           Target  TargetValue  
0  ConfirmedCases            0  
1      Fatalities            0  
2  ConfirmedCases            0  
3      Fatalities            0  
4  ConfirmedCases            0

In [7]:
#show shape
df.shape

(969640, 9)

In [8]:
#show info for dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 969640 entries, 0 to 969639
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Id              969640 non-null  int64  
 1   County          880040 non-null  object 
 2   Province_State  917280 non-null  object 
 3   Country_Region  969640 non-null  object 
 4   Population      969640 non-null  int64  
 5   Weight          969640 non-null  float64
 6   Date            969640 non-null  object 
 7   Target          969640 non-null  object 
 8   TargetValue     969640 non-null  int64  
dtypes: float64(1), int64(3), object(5)
memory usage: 66.6+ MB


There are Missing Value in `Country` and `Province_State` Variabel.

In [9]:
#show missing value 
df.isnull().sum()

Id                    0
County            89600
Province_State    52360
Country_Region        0
Population            0
Weight                0
Date                  0
Target                0
TargetValue           0
dtype: int64

There are 89600 Missing Value in `Country` and 52360 in `Province_State` Variabel.

In [10]:
#show unique each columns
for i in df.columns:
  print(f'Unique_{i}')
  print(df[i].value_counts())
  print("")

Unique_Id
1         1
646431    1
646419    1
646420    1
646421    1
         ..
323219    1
323220    1
323221    1
323222    1
969640    1
Name: Id, Length: 969640, dtype: int64

Unique_County
Washington    8680
Jefferson     7280
Franklin      7000
Lincoln       6720
Jackson       6720
              ... 
Harford        280
Garrett        280
Charles        280
Cecil          280
Weston         280
Name: County, Length: 1840, dtype: int64

Unique_Province_State
Texas                       71400
Georgia                     44800
Virginia                    37520
Kentucky                    33880
Missouri                    32760
                            ...  
Liaoning                      280
Jilin                         280
Jiangxi                       280
Jiangsu                       280
Turks and Caicos Islands      280
Name: Province_State, Length: 133, dtype: int64

Unique_Country_Region
US                895440
China               9520
Canada              3640
United King

Count of unique in `County`,`Provine_State`,`Country_Region` and `Population` are not same. So, didn't know categories for `Population` variabel. Maybe, because effect of the Missing Value or else.

In [11]:
#show all data
df

Id County Province_State Country_Region  Population    Weight  \
0            1    NaN            NaN    Afghanistan    27657145  0.058359   
1            2    NaN            NaN    Afghanistan    27657145  0.583587   
2            3    NaN            NaN    Afghanistan    27657145  0.058359   
3            4    NaN            NaN    Afghanistan    27657145  0.583587   
4            5    NaN            NaN    Afghanistan    27657145  0.058359   
...        ...    ...            ...            ...         ...       ...   
969635  969636    NaN            NaN       Zimbabwe    14240168  0.607106   
969636  969637    NaN            NaN       Zimbabwe    14240168  0.060711   
969637  969638    NaN            NaN       Zimbabwe    14240168  0.607106   
969638  969639    NaN            NaN       Zimbabwe    14240168  0.060711   
969639  969640    NaN            NaN       Zimbabwe    14240168  0.607106   

              Date          Target  TargetValue  
0       2020-01-23  ConfirmedCases            0  
1       2020-01-23      Fatalities            0  
2       2020-01-24  ConfirmedCases            0  
3       2020-01-24      Fatalities            0  
4       2020-01-25  ConfirmedCases            0  
...            ...             ...          ...  
969635  2020-06-08      Fatalities            0  
969636  2020-06-09  ConfirmedCases           27  
969637  2020-06-09      Fatalities            0  
969638  2020-06-10  ConfirmedCases            6  
969639  2020-06-10      Fatalities            0  

[969640 rows x 9 columns]

Based on the dataset, didn't known spesifik categories for each unique in `Population` variabel. Because there are so many Missing Value in `County` and `Province_State`. But, we still use `Country_Region` variabel as categori for `Population` although there are `Country_Region` has a many unique `Population` because in that `Country_Region` has a many `County` or ` Province_State`.

###2.Cleaning & Handling Dataframe

In [12]:
#drop irrelevant columns
df=df.drop(columns=['Id'])
df.columns

Index(['County', 'Province_State', 'Country_Region', 'Population', 'Weight',
       'Date', 'Target', 'TargetValue'],
      dtype='object')

In [13]:
#Missing Value Handling
df = df.drop(['County','Province_State'], axis=1)
print(df.columns)
print(df.isnull().sum())
df.head()

Index(['Country_Region', 'Population', 'Weight', 'Date', 'Target',
       'TargetValue'],
      dtype='object')
Country_Region    0
Population        0
Weight            0
Date              0
Target            0
TargetValue       0
dtype: int64


Country_Region  Population    Weight        Date          Target  \
0    Afghanistan    27657145  0.058359  2020-01-23  ConfirmedCases   
1    Afghanistan    27657145  0.583587  2020-01-23      Fatalities   
2    Afghanistan    27657145  0.058359  2020-01-24  ConfirmedCases   
3    Afghanistan    27657145  0.583587  2020-01-24      Fatalities   
4    Afghanistan    27657145  0.058359  2020-01-25  ConfirmedCases   

   TargetValue  
0            0  
1            0  
2            0  
3            0  
4            0

In [14]:
#Show Unique each column
for i in df.columns:
  print(f'Unique_{i}')
  print(df[i].value_counts())
  print("")

Unique_Country_Region
US                895440
China               9520
Canada              3640
United Kingdom      3080
France              3080
                   ...  
Ghana                280
Greece               280
Grenada              280
Guatemala            280
Zimbabwe             280
Name: Country_Region, Length: 187, dtype: int64

Unique_Population
14651       840
2530        840
2356        560
19983       560
2013        560
           ... 
9564        280
66911       280
2648        280
3250        280
14240168    280
Name: Population, Length: 3408, dtype: int64

Unique_Weight
1.276101    420
0.104250    420
1.042499    420
0.127610    420
1.287806    280
           ... 
0.111490    140
1.061230    140
0.106123    140
0.668232    140
0.607106    140
Name: Weight, Length: 6816, dtype: int64

Unique_Date
2020-01-23    6926
2020-04-28    6926
2020-04-22    6926
2020-04-23    6926
2020-04-24    6926
              ... 
2020-03-03    6926
2020-03-02    6926
2020-03-01    6926

###3.Data Manipulation

####a. Change Data Structure

In [15]:
#show data structure
df

Country_Region  Population    Weight        Date          Target  \
0         Afghanistan    27657145  0.058359  2020-01-23  ConfirmedCases   
1         Afghanistan    27657145  0.583587  2020-01-23      Fatalities   
2         Afghanistan    27657145  0.058359  2020-01-24  ConfirmedCases   
3         Afghanistan    27657145  0.583587  2020-01-24      Fatalities   
4         Afghanistan    27657145  0.058359  2020-01-25  ConfirmedCases   
...               ...         ...       ...         ...             ...   
969635       Zimbabwe    14240168  0.607106  2020-06-08      Fatalities   
969636       Zimbabwe    14240168  0.060711  2020-06-09  ConfirmedCases   
969637       Zimbabwe    14240168  0.607106  2020-06-09      Fatalities   
969638       Zimbabwe    14240168  0.060711  2020-06-10  ConfirmedCases   
969639       Zimbabwe    14240168  0.607106  2020-06-10      Fatalities   

        TargetValue  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
...             ...  
969635            0  
969636           27  
969637            0  
969638            6  
969639            0  

[969640 rows x 6 columns]

Because the structure of dataframe is not efficient to use for analysis. We can repair that structur of dataframe.

In [16]:
#repair dataframe structure
df_cc = df[df['Target'].isin(['ConfirmedCases'])] #get data with filtering for `Target` variabel is 'ConfirmedCases'
df_cc = df_cc.drop(columns=['Target']) #drop `Target` variabel to repair be column
df_cc = df_cc.reset_index(drop=True) #make new index for dataframe has filtered
df_cc.rename(columns={"TargetValue":"Confirmed Cases"}, inplace=True) #rename `TargetValue` to be show count of `Confirmed Cases`
print(df_cc.shape) #show dataframe shape
df_cc.head() #show dataframe head

(484820, 5)


Country_Region  Population    Weight        Date  Confirmed Cases
0    Afghanistan    27657145  0.058359  2020-01-23                0
1    Afghanistan    27657145  0.058359  2020-01-24                0
2    Afghanistan    27657145  0.058359  2020-01-25                0
3    Afghanistan    27657145  0.058359  2020-01-26                0
4    Afghanistan    27657145  0.058359  2020-01-27                0

In [17]:
#repair dataframe structure
df_f = df[df['Target'].isin(['Fatalities'])] #get data with filtering for `Target` variabel is 'FatalitiesCases'
df_f = df_f.drop(columns=['Target']) #drop `Target` variabel to repair be column
df_f = df_f.reset_index(drop=True) #make new index for dataframe has filtered
df_f.rename(columns={"TargetValue":"Fatalities Cases"}, inplace=True) #rename `TargetValue` to be show count of `Fatalities Cases`
print(df_f.shape) #show dataframe shape
df_f.head() #show dataframe head

(484820, 5)


Country_Region  Population    Weight        Date  Fatalities Cases
0    Afghanistan    27657145  0.583587  2020-01-23                 0
1    Afghanistan    27657145  0.583587  2020-01-24                 0
2    Afghanistan    27657145  0.583587  2020-01-25                 0
3    Afghanistan    27657145  0.583587  2020-01-26                 0
4    Afghanistan    27657145  0.583587  2020-01-27                 0

In [18]:
#repair dataframe structure
cc_value = df_cc['Confirmed Cases'] #get 'Confirmed Cases' in df_cc dataframe
df_f['Confirmed Cases'] = cc_value #make new column in df_f dataframe filled the 'Confirmed Cases' value
print(df_f.shape) #show df_f dataframe has repaired
df_f #show df_f dataframe

(484820, 6)


Country_Region  Population    Weight        Date  Fatalities Cases  \
0         Afghanistan    27657145  0.583587  2020-01-23                 0   
1         Afghanistan    27657145  0.583587  2020-01-24                 0   
2         Afghanistan    27657145  0.583587  2020-01-25                 0   
3         Afghanistan    27657145  0.583587  2020-01-26                 0   
4         Afghanistan    27657145  0.583587  2020-01-27                 0   
...               ...         ...       ...         ...               ...   
484815       Zimbabwe    14240168  0.607106  2020-06-06                 0   
484816       Zimbabwe    14240168  0.607106  2020-06-07                 0   
484817       Zimbabwe    14240168  0.607106  2020-06-08                 0   
484818       Zimbabwe    14240168  0.607106  2020-06-09                 0   
484819       Zimbabwe    14240168  0.607106  2020-06-10                 0   

        Confirmed Cases  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
484815               14  
484816                3  
484817                5  
484818               27  
484819                6  

[484820 rows x 6 columns]

The new structure of dataframe make the dataframe has a new count of rows. dataframe has 484820 rows now.

In [19]:
#download result data manipulation
df_f.to_csv(index=False)
compression_opts = dict(method='zip',
                        archive_name='global_covid.xlsx')  
df_f.to_csv('out.zip', index=False,
          compression=compression_opts) 
from pathlib import Path  
filepath = Path('/content/drive/MyDrive/Colab Notebooks/global_covid.xlsx')  
filepath.parent.mkdir(parents=True, exist_ok=True)  
df_f.to_csv(filepath)  
import os  
os.makedirs('/content/drive/MyDrive/Colab Notebooks/', exist_ok=True)  
df_f.to_csv('/content/drive/MyDrive/Colab Notebooks/global_covid.xlsx')  

The dataframe has downloaded in the local drive.

####b. Aggregation Manipulation For Looker

Because looker studio can not use sum_distict and `Population` variabel has a same value in record. So, to show the count of `Population` we have use different way. One way, we can use average function. So, we have known that frequently of duplicated value.

In [20]:
#show value count to know minimum looping for population and country_ragion variabel
print(df_f['Population'].value_counts())
print(df_f['Country_Region'].value_counts())

14651       420
2530        420
2356        280
19983       280
2013        280
           ... 
9564        140
66911       140
2648        140
3250        140
14240168    140
Name: Population, Length: 3408, dtype: int64
US                447720
China               4760
Canada              1820
United Kingdom      1540
France              1540
                   ...  
Ghana                140
Greece               140
Grenada              140
Guatemala            140
Zimbabwe             140
Name: Country_Region, Length: 187, dtype: int64


Based on above output, we known that each unique has 140 duplicated record which is value_counts of `Date` variabel. So, we use 140 in counting real `Population`.

In [27]:
#show sum population variabel for each country_region
agg2 = df_f.groupby('Country_Region')['Population'].sum()/140
agg2
agg3 = pd.DataFrame(agg2)
agg3['Population'].sum()

9419799780.0

Based on above output, we known that counting of `Population` used the `Average` way is usefull because result rigth value for show sum of all `Population`.